### PR TITLE
Dev/feature yolo

### PR DESCRIPTION
**Pull request by bamsumit**
Issue Number: 242

Objective of pull request:
YOLO SDNN training and GPU inference pipeline

## Pull request checklist

Your PR fulfills the following requirements:
- [x] [Issue](https://github.com/lava-nc/lava-dl/issues) created that explains the change and why it's needed
- [ ] Tests are part of the PR (for bug fixes / features)
- [x] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [x] Build tests (`pytest`) passes locally

## Pull request type

 

 

Please check your PR type:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/lava/lib/dl/slayer/__init__.py [hunk 1]
```diff
@@ -12,6 +12,7 @@
 from . import io
 from . import auto
 from . import utils
+from . import object_detection
```
Comment (bamsumit) [inline src/lava/lib/dl/slayer/__init__.py:15]: Note: Try and find an abbreviation for object detection

File: src/lava/lib/dl/slayer/object_detection/boundingbox/utils.py [hunk 2]
```diff
@@ -0,0 +1,595 @@
...
+    ann['annotation']['size']['width'] = width
+    for i in range(len(ann['annotation']['object'])):
+        xmin = int(ann['annotation']['object'][i]['bndbox']['xmin'])
+        xmax = int(ann['annotation']['object'][i]['bndbox']['xmax'])
+        ann['annotation']['object'][i]['bndbox']['xmin'] = width_fx(xmax)
+        ann['annotation']['object'][i]['bndbox']['xmax'] = width_fx(xmin)
+    return annotation
+
+
+def create_video(inputs, targets, predictions, video_output_path, BOX_COLOR_MAP, classes):
```
Comment (bamsumit) [inline src/lava/lib/dl/slayer/object_detection/boundingbox/utils.py:569]: Note: Missing typehints, proper formatting and docstring

File: tutorials/lava/lib/dl/slayer/tiny_yolo_sdnn/trained/network.pt
Comment (bamsumit) [inline tutorials/lava/lib/dl/slayer/tiny_yolo_sdnn/trained/network.pt:1]: @elrond91 please update the final trained results and network description from @timcheck's experiments.

**Comment by bamsumit** [comment]
Dear Reviewers,

Do not fret seeing a lot of files in the PR. Here are the main files to be reviewed in the order of priority. Please :
* `tutorials/lava/lib/dl/slayer/tiny_yolo_sdnn/README.md`
* `tutorials/lava/lib/dl/slayer/tiny_yolo_sdnn/train_sdnn.py`
* `tutorials/lava/lib/dl/slayer/tiny_yolo_sdnn/inference.ipynb`
* `src/lava/lib/dl/slayer/object_detection/models/tiny_yolov3_str.py`
* `src/lava/lib/dl/slayer/object_detection/models/yolo_kp.py`
* `src/lava/lib/dl/slayer/object_detection/boundingbox/metrics.py`
* `src/lava/lib/dl/slayer/object_detection/boundingbox/utils.py`
* `src/lava/lib/dl/slayer/object_detection/dataset/bdd100k.py`
* `src/lava/lib/dl/slayer/loss.py`
* and the rest :)

File: src/lava/lib/dl/slayer/object_detection/boundingbox/utils.py [hunk 3]
```diff
@@ -0,0 +1,694 @@
...
+                    ious * label_match > nms_threshold
+                ).long().triu(1).sum(dim=0,
+                                     keepdim=True).T.expand_as(detections) == 0
+
+                detections = detections[keep].reshape(-1, 6).contiguous()
+                if detections.shape[0] == prev_objects:
+                    break
+                else:
+                    prev_objects = detections.shape[0]
+        # #
```
Comment (mgkwill) [inline src/lava/lib/dl/slayer/object_detection/boundingbox/utils.py:97]: Remove commented code.

File: src/lava/lib/dl/slayer/object_detection/dataset/bdd100k.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    str
+        String without the suffix.
+    """
+    if suffix and input_string.endswith(suffix):
+        return input_string[:-len(suffix)]
+    return input_string
+
+
+class _BDD(Dataset):
+    # Low level BDD100K dataset. To be wrapped around, do not use extrnally.
```
Comment (mgkwill) [inline src/lava/lib/dl/slayer/object_detection/dataset/bdd100k.py:42]: Slight typo in `extrnally`.

File: src/lava/lib/dl/slayer/object_detection/models/Trained_tiny_yolov3_str/.gitattributes [hunk 5]
```diff
@@ -0,0 +1 @@
+network.pt filter=lfs diff=lfs merge=lfs -text
```
Comment (mgkwill) [inline src/lava/lib/dl/slayer/object_detection/models/Trained_tiny_yolov3_str/.gitattributes:1]: Thanks for adding these to Git LFS!

File: src/lava/lib/dl/slayer/object_detection/yolo_base.py [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+        print(f'{torch.isinf(width).any()=}')
+        print(f'{torch.isnan(height).any()=}')
+        print(f'{torch.isinf(height).any()=}')
+        raise RuntimeError('Ecountered NaN and Inf!')
+
+    return x  # batch, anchor, height, width, predictions, time
+
+
+def _yolo_target(x: torch.tensor, anchors: torch.tensor) -> torch.tensor:
+    # Generates target tesnsor for a given anchor resolution.
```
Comment (mgkwill) [inline src/lava/lib/dl/slayer/object_detection/yolo_base.py:59]: Slight typo in comment `tesnsor`

File: src/lava/lib/dl/slayer/object_detection/yolo_base.py [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+        -------
+        torch.tensor
+            Loss value.
+        """
+        loss = 0
+        loss_distr = []
+        # for time in range(len(targets)):
+        for time in range(predictions[0].shape[-1]):
+            prediction = [p[..., time].clone() for p in predictions]
+            # target = self.build_targets(prediction, targets[time])
```
Comment (mgkwill) [inline src/lava/lib/dl/slayer/object_detection/yolo_base.py:287]: Remove commented code.

**Review by mgkwill — APPROVED** [review]
A few places to remove commented code and a couple of comment spelling errors, else LGTM.
